plugins/elantp/fu-elantp-hid-haptic-device.c

```diff
@@ -630,36 +629,38 @@ fu_elantp_hid_haptic_device_write_firmware(FuDevice *device,
 							error))
 				return FALSE;
 
-		if (!fu_elantp_hid_haptic_device_ensure_eeprom_iap_ctrl(FU_DEVICE(parent), self, &ctrl_result, error)) 
-			return FALSE;	
-
-		if (ctrl_result == 1) {
-			retry_cnt = 0;	
-
-			/* update progress */
-			checksum += csum_tmp;
-			fu_progress_set_percentage_full(fu_progress_get_child(progress),
-							(gsize)i + 1,
-							(gsize)chunks->len + 1);
-		} else if (ctrl_result == 0) {
-			i -= 1;
-			retry_cnt++;
-			if (retry_cnt >= 3) {
+		if (!fu_elantp_hid_haptic_device_ensure_eeprom_iap_ctrl(FU_DEVICE(parent), self, &error_iapctrl)) {		
+			if (g_error_matches(error_iapctrl, G_IO_ERROR, G_IO_ERROR_BUSY)) {
+				i -= 1;
+				retry_cnt++;
+				if (retry_cnt >= 3) {
+					g_set_error(error,
+						FWUPD_ERROR,
+						FWUPD_ERROR_WRITE,
+						"bootloader reports failed write: 0x%x (%s)",
+						self->iap_ctrl,
+						error_iapctrl->message);
+					return FALSE;
+				}
+				break;
+			} else {
```

Comment:
i don't think you need this `else`

Reply:
If the error is G_IO_ERROR_BUSY, will retry, but other failures will return FALSE.